### PR TITLE
kubelet register: update new labels to existing node

### LINKED ISSUE #28051: Restarting kubelet with new node labels doesn't update labels for node
kubelet version: 1.2.4

Steps to reproduce:
- start kubelet with some labels using the `--node-labels` flag
- wait for the node to register itself with the master
- restart kubelet with a different set of labels

Expected result:
`kubectl get no --show-labels` shows the node with the updated labels

Actual result:
`kubectl get no --show-labels` shows the node with the original labels

### PR DESCRIPTION
**Pull request by resouer**
Fixes: #28051 , kubelet should be able to apply new labels specified by `--node-labels` after restarted.

The reconcile strategy in this PR is: always use new label to replace the old labels.

### CODE REVIEW

**Comment by resouer** [comment]
@vishh any time to take a look at this fix?

**Comment by vishh** [comment]
I recall specifically disabling updates to node labels because the concept of `owners` is unclear for node labels. A user or a k8s component or kubelet can apply node labels and they can step on each others toes. 
The priority order is also undefined. 
With this PR you are essentially making kubelet win over other sources. This might break some users. This is especially a problem while removing labels. If a user removes a label originally specified as part of kubelet startup, then how will kubelet be able to identify the missing/removed label and discern if it is safe to remove that label?

**Comment by resouer** [comment]
@vishh In my opinion, `--node-labels` is expected to win over other sources and that's why it is configured directly to kubelet binary. It should reset values in `--node-labels` which have been modified by user during restart. 

If a user want to remove any existing values in `--node-labels` permanently, he should modify the this configuration.

@timothysc Do rh folks expect the same behaviour? Or we do have other reasons to reconcile user's modification with `--node-labels`. Personally, it is so weird to preserve users change while keep `--node-labels` with some outdated value ...

**Comment by timothysc** [comment]
@resouer I think the aim was to moved towards storing that data in etcd and making that the canonical source of truth vs. command line options.  

/cc @kubernetes/sig-node-misc

**Comment by euank** [comment]
Related, https://github.com/kubernetes/kubernetes/pull/29459 (storing kubelet config in a configmap instead of CLI).

I do think re-applying labels the user removed is confusing. I think that the following series of changes might be better:

1. kubelet config, including labels, is in a configmap
2. kubelet watchis its own configmap to dynamically reconfigure itself, specifically for node-label changes
3. kubectl flags related to adding and removing labels mutate the configmap instead of the node object directly (or in addition to / right before updating the node object).

As-is, I do think this is a scary change.

At the very least, it needs a release-note and warning associated with it.

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -155,6 +160,22 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *v1.Node) bool {
 	return false
 }
 
+// reconcileLablesWithExistingNode reconciles labels on a new node and the existing node, returning
+// whether the existing node must be updated.
+// Current reconcile strategy: always use labels on new node.
+func (kl *Kubelet) reconcileLablesWithExistingNode(node, existingNode *v1.Node) bool {
```
Comment (euank) [inline pkg/kubelet/kubelet_node_status.go]: This function name is misleading. There's no reconciliation here.

The function name could more accurately be called "clobberAnyUserConfigurationChanges". It's not reconciling.

File: pkg/kubelet/kubelet_node_status.go [hunk 2]
```diff
@@ -155,6 +160,22 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *v1.Node) bool {
...
+	existingLabels := existingNode.GetLabels()
+	newLabels := node.GetLabels()
+
+	// always use newly provided labels to update the existing node
+	if !reflect.DeepEqual(existingLabels, newLabels) {
```
Comment (euank) [inline pkg/kubelet/kubelet_node_status.go]: please don't use reflect.DeepEqual outside of test code. It's slow and icky.

**Review by euank — CHANGES_REQUESTED** [review]
I think that this change is incomplete.

If we want the kubelet to be the source of truth for node labels, I think we also need to make the "kubectl label" command error out when applied to nodes (since after-all if the kubelet crashes, updates, or otherwise restarts at any point in the future the command will have had no effect).

**Comment by resouer** [comment]
> "kubectl label" command error out when applied to nodes 

Yes, this is needed when existing label is being updated.

@euank Is your suggestion has been all covered by #29459 ? Sorry for missing that proposal, but I definitely prefer dynamic configuration too if it's possible.

**Comment by euank** [comment]
@resouer I'm being facetious about the "make the `kubectl label` command error..." bit. What I meant was that I don't think this change is reasonable from a UX perspective.

The "dynamic config" proposal does not cover labels at all, but I do think that building label changes on top of it makes it possible to create better behaviour.

**Comment by resouer** [comment]
I choose close this now and watch https://github.com/kubernetes/kubernetes/pull/29459

Thanks for comments and let's try to solve it from another side.

**Comment by SnowmanCharles** [comment]
Looks like the current behavior is that kubelet still doesn't update labels. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L109-L125. 

Although I agree that user labels usually are more important, I feel like we still need a way to enforce kubelet to apply some labels. Is that supported by dynamic config or something else? If not, can we have kubelet expose some simple flags? For example, --ForceAdd, --ForcePatch, --ForceCleanOld.

**Pull request closed without merging**